drivers: modem: generalize Quectel BG9X support

guildeol-stra opened the pull request:
Replace specific reference for Quectel BG95 in Modem Cellular with
BG9x, since the setup code should apply for other members of this
family.

Rename the 'quectel,bg95' DTS binding to 'quectel,bg9x' and add
optional 'mdm-reset-gpios' property.

To avoid confusion, delete the old 'quectel,bg9x' DTS binding and its
implementation.

Fixes #93115

guildeol-stra commented:
> > Hi, @rerickson1
> > I think the LTS versions could benefit from this. Can this also be backported?
> 
> Does this issue exist in 3.7?

Yes. We use 3.7 as our base version and we first identified the issue there.

guildeol-stra commented:
> > > Hi, @rerickson1
> > > I think the LTS versions could benefit from this. Can this also be backported?
> > 
> > 
> > Does this issue exist in 3.7?
> 
> Can you please writeup a issue quick so it can be linked to this PR? This is the official process that should be followed when including a fix like this.

Done. Added the link and also ammended my commit message because this did not cause a compilation issue, just disabled the config implicitly.

rerickson1 commented:
> > > > Hi, @rerickson1
> > > > I think the LTS versions could benefit from this. Can this also be backported?
> > > 
> > > 
> > > Does this issue exist in 3.7?
> > 
> > 
> > Can you please writeup a issue quick so it can be linked to this PR? This is the official process that should be followed when including a fix like this.
> 
> Done. Added the link and also amended my commit message because this did not cause a compilation issue, just disabled the config implicitly.
> 
> Please let me know if there's anything else needed.

Looks good. We'll see if its possible to get this bug fix in for 4.2. If not, it will come after.

bjarki-andreasen reviewed (changes requested):
That was by design :) the bg9x compat is a relic used only by the quectel-bg9x.c driver, modem_cellular (and zephyr in general) requires the devicetree to have specific device compats in the devicetree, like quectel,bg95 (which is why it was added and is used by the modem subsystem)

To elaborate a bit, `quectel,bg9x` is a malformed devicetree compatible, not a driver :) quectel-bg9x.c and modem_ceullar.c are drivers. These support specific hardware, which is what devicetree compatibles specify. If you have a bg95, or bg96 (something matching the "wildcard" bg9x), that compatible has to be added to zephyr, and the modem_cellular.c driver expanded to support it :)

rerickson1 reviewed (changes requested):
The `quectel-bg9x.c` driver cannot be removed without first deprecating it.
Please see: https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html

guildeol-stra committed:
drivers: modem: deprecate MODEM_QUECTEL_BG9X

The modem_cellular implementation should suffice to handle the base
usage of generic BG9x family devices, so deprecate the quectel-bg9x.c
implementation.

Signed-off-by: Guilherme Costa <guilhermecosta@stratioautomotive.com>

dsseng commented on dts/bindings/modem/quectel,bg95.yaml:
As one of the previous reviewers wrote, please do not remove existing drivers and bindings without going through a deprecation process, they are very likely relied upon by downstream DTs

dsseng commented on dts/bindings/modem/quectel,bg9x.yaml:
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) 2020 Analog Life LLC
+# Copyright (c) 2023 Bjarki Arge Andreasen
```
Please keep the current author information and add your own data besides, with the date of the change

```suggestion
# Copyright (c) 2020 Analog Life LLC
# Copyright (c) 2025 Bjarki Arge Andreasen
```

dsseng commented on dts/bindings/modem/quectel,bg9x.yaml:
```diff
@@ -10,7 +10,6 @@ include: uart-device.yaml
 properties:
   mdm-power-gpios:
     type: phandle-array
-    required: true
```
If removing a validation from here, please make sure consumers (the driver) check for this property e.g. using assertions

dsseng commented on dts/bindings/modem/quectel,bg9x.yaml:
```diff
@@ -20,3 +19,11 @@ properties:
   mdm-wdisable-gpios:
     type: phandle-array
+
+  autostarts:
```
`always-on`? Just an idea

dsseng replied:
You should probably keep this addition, it seems useful. Also if it matches names in source, that name is good, keep it

dsseng commented on samples/net/cellular_modem/boards/b_u585i_iot02a.overlay:
```diff
@@ -31,7 +31,7 @@
 
 	modem: modem {
-		compatible = "quectel,bg95";
+		compatible = "quectel,bg9x";
```
Have you tested this change on this board?

guildeol-stra replied:
Reverted this

dsseng commented on drivers/modem/modem_cellular.c:
```diff
@@ -540,7 +540,7 @@ MODEM_CHAT_MATCHES_DEFINE(dial_abort_matches,
 #if DT_HAS_COMPAT_STATUS_OKAY(swir_hl7800) || DT_HAS_COMPAT_STATUS_OKAY(sqn_gm02s) ||              \
 	DT_HAS_COMPAT_STATUS_OKAY(quectel_eg800q) || DT_HAS_COMPAT_STATUS_OKAY(quectel_eg25_g) ||  \
-	DT_HAS_COMPAT_STATUS_OKAY(quectel_bg95) || DT_HAS_COMPAT_STATUS_OKAY(simcom_a76xx)
+	DT_HAS_COMPAT_STATUS_OKAY(quectel_bg9x) || DT_HAS_COMPAT_STATUS_OKAY(simcom_a76xx)
```
bg95 compatible should not be removed without going through a deprecation process, please handle them side-by-side, just use `||` in most places

guildeol-stra replied:
Done. I modified the places that originally had only `DT_HAS_COMPAT_STATUS_OKAY(quectel_bg95)` to accept both BG95 and BG9X compatibles, but renamed the common scripts to bg9x only, as I believe that is more sensible.

dsseng commented on drivers/modem/modem_cellular.c:
```diff
@@ -538,9 +538,10 @@ MODEM_CHAT_MATCHES_DEFINE(dial_abort_matches,
-#if DT_HAS_COMPAT_STATUS_OKAY(swir_hl7800) || DT_HAS_COMPAT_STATUS_OKAY(sqn_gm02s) ||              \
-	DT_HAS_COMPAT_STATUS_OKAY(quectel_eg800q) || DT_HAS_COMPAT_STATUS_OKAY(quectel_eg25_g) ||  \
-	DT_HAS_COMPAT_STATUS_OKAY(quectel_bg95) || DT_HAS_COMPAT_STATUS_OKAY(simcom_a76xx)
+#if DT_HAS_COMPAT_STATUS_OKAY(swir_hl7800) || DT_HAS_COMPAT_STATUS_OKAY(sqn_gm02s) || \
+	DT_HAS_COMPAT_STATUS_OKAY(quectel_eg800q) || DT_HAS_COMPAT_STATUS_OKAY(quectel_eg25_g) || \
```
I think formatting changes here might be out of scope and I'm not sure if there actually is a general rule regarding spaces before \

guildeol-stra replied:
I tried keeping the same "style" as the other.  This file raises a few formatting warnins when running the compilance checks, but I think they would take the focus away from the other changes, so I decided to leave it as is.

guildeol-stra commented:
ping @rerickson1, @bjarki-andreasen

bjarki-andreasen reviewed (changes requested):
Given BG9X is not an actual modem which exists out there, the compat `quectel,bg9x.yaml` should be deprecated along with the BG9X driver. Modem cellular already supports the bg95, I'm expecting the reason bg9x is being used by anyone is that they have a bg96? how about adding `quectel,bg96.yaml` and adding that to modem_cellular?

guildeol-stra committed:
dts: bindings: Add 'quectel,bg96'

Add a compatible for Quectel BG96 to be used in the modem_cellular
subsystem.

Signed-off-by: Guilherme Costa <guilhermecosta@stratioautomotive.com>

guildeol-stra committed:
drivers: modem: Add support for Quectel BG96

Added support for compatible 'quectel,bg96' in modem_cellular.c

Renamed BG95-specific variables to BG9x, since the initialization
for BG95 and BG9x is similar.

Signed-off-by: Guilherme Costa <guilhermecosta@stratioautomotive.com>

bjarki-andreasen commented:
> > Given BG9X is not an actual modem which exists out there, the compat `quectel,bg9x.yaml` should be deprecated along with the BG9X driver. Modem cellular already supports the bg95, I'm expecting the reason bg9x is being used by anyone is that they have a bg96? how about adding `quectel,bg96.yaml` and adding that to modem_cellular?
> 
> Sure thing. Should I keep the chat scripts rename to `*_bg9x` in `modem_cellular.c`? For example, `quectel_bg95_init_chat_script` becomes `quectel_bg9x_init_chat_script`. I feel like it makes sense, at least while there is no reason for creating separate scripts and leaving it as BG95 would be confusing.

Yes, inside the driver, the wildcard makes sense :)

bjarki-andreasen reviewed (approved):
Looks good!
